### PR TITLE
Proof of concept to handle issue with JS not initialized on start to fix #107

### LINKED ISSUE #107: [iOS 13] callKit displayed from AppDelegate.m before incoming call listeners
After updating to 3.0.3 and updating my code to meet iOS 13 requirements so CallKit is reported from AppDelegate. But the problem is that when app is closed or in background, the callKit displays very fast before my call class get loaded. So the callKit is shown on screen before listeners are ready and if I accept or reject the call immediately nothing happened. I need to wait for 3 seconds until call listeners are ready then I can accept or reject.

How can I delay the callKit in AppDelegate until my call class is fully loaded?

### PR DESCRIPTION
**Pull request by konradkierus**
Proof of concept to handle issue with JS not initialized on start to fix #107

Described in details here: https://github.com/react-native-webrtc/react-native-callkeep/issues/107#issuecomment-581845376

Example usage:
```
import { each, compact, last, get } from 'lodash';

RNCallKeep.addEventListener(
  'didLoadWithEvents',
  this.onDidLoadWithNativeEvents
);

onDidLoadWithNativeEvents(events) {
  const validEvents = compact(events);
  let callDataToAdd = null;
  let callDataToAnswer = null;
  let callDataToReject = null;
  let callDataToInitiateCall = null;
  each(compact(validEvents), event => {
    const { name, data } = event;
    if (name === 'RNCallKeepDidDisplayIncomingCall') {
      callDataToAdd = data;
      callDataToAnswer = null;
      callDataToReject = null;
    }
    if (name === 'RNCallKeepPerformAnswerCallAction') {
      callDataToReject = null;
      callDataToAnswer = data;
    }
    if (name === 'RNCallKeepPerformEndCallAction') {
      callDataToAnswer = null;
      callDataToReject = data;
    }
  });

  const lastEventName = get(last(validEvents), 'name');
  if (lastEventName === 'RNCallKeepDidReceiveStartCallAction') {
    callDataToInitiateCall = get(last(validEvents), 'data');
    callDataToAnswer = null;
    if (!callDataToReject) {
      callDataToAdd = null;
    }
  }

  if (callDataToAdd) {
    this.onDisplayIncomingCall(callDataToAdd);
    if (callDataToReject) {
      this.onEndCall(callDataToReject);
    }
    if (callDataToAnswer) {
      this.onAnswerCall(callDataToAnswer);
    }
  }
  if (callDataToInitiateCall) {
    this.onStartCall(callDataToAnswer);
  }
}
```

### CODE REVIEW

**Comment by manuquentin** [comment]
Thanks @konradkierus I'll check it. Can you also update the documentation ?

**Comment by konradkierus** [comment]
@manuquentin just did, feel free to modify if you think it needs more details

File: ios/RNCallKeep/RNCallKeep.m [hunk 1]
```diff
@@ -92,10 +97,36 @@ - (void)dealloc
         RNCallKeepPerformPlayDTMFCallAction,
         RNCallKeepDidToggleHoldAction,
         RNCallKeepProviderReset,
-        RNCallKeepCheckReachability
+        RNCallKeepCheckReachability,
+        RNCallKeepDidLoadWithEvents
     ];
 }
 
+- (void)startObserving
```
Comment (zxcpoiu) [inline ios/RNCallKeep/RNCallKeep.m]: where who to call the `startObserving`?

**Comment by zxcpoiu** [comment]
@konradkierus 
Thanks for your work :+1: 

I have some thoughts for open discussion.

The principle is: event order is important.
And we hope everything happens behind the scene, so user won't aware and don't have to do additional things.

Based on your concept, how about:

1. When RN Bridge did not up, queue up events ( the same ) 
2. When user setup all required event listener on JS Side, call a function ( like `RNCallKeep.ready()` )
   ( or at some point, we can call `this.ready()` somewhere in RNCallKeep js side? constructor? )
3. In Native Side `ready()` function, flush and fire all queued events ( which is in order ), then set `hasListener = YES` ( or just `ready = YES` )

we might need to  give user the last events, whether store in a variable, then let user to `getLastEvent()` or just return then last event in `ready()` function.

### should discuss:

* Should events have to expire?
* Should some events override each other? like if EndCall first, then AsnwerCall the second event in the same uuid:
  - should we only store EndCall and ignore AnswerCall
  - or should we only store AnswerCall and remove EndCall event
  - or should we just send two events in order

File: ios/RNCallKeep/RNCallKeep.m (hunk 1 above)
Comment (konradkierus) [reply ios/RNCallKeep/RNCallKeep.m]: @zxcpoiu React Native calls this method internally for classes which inherit from RCTEventEmitter, see here: https://github.com/facebook/react-native/blob/d0871d0a9a373e1d3ac35da46c85c0d0e793116d/React/Modules/RCTEventEmitter.m#L99 and here: https://reactnative.dev/docs/native-modules-ios#optimizing-for-zero-listeners

**Comment by konradkierus** [comment]
@zxcpoiu 

I don't see issues with implementing `didLoadWithEvents` event internally in the JS part of the module and then call `ready()` or something similar.
Unfortunately, I don't currently work on the project which was using this approach so I won't be able to test any further changes I would do.

Now, regarding your questions. When I've started thinking about solution for the #107 issue, I was asking myself very similar questions which you are now asking me.
I've decided that it's very hard to implement one and correct solution for everyone as people may have different requirements. Some people (like me) would like to have full control on events order. Other people may be interested about last event only. That's why I've implemented it in this way so anyone could decide by themselves.

File: ios/RNCallKeep/RNCallKeep.m (hunk 1 above)
Comment (zxcpoiu) [reply ios/RNCallKeep/RNCallKeep.m]: ah, I just missed this part, thank you！

**Comment by zxcpoiu** [comment]
Agreed, it's hard to implement a solution suit well for all cases.

And yes, if the js part called internally in js module side, the user won't bother much. That's my whole point try to avoid add additional handling.

That's wait others thoughts and different use cars.

Thanks for your elaboration.

**Comment by manuquentin** [comment]
Hi @konradkierus,
I've created the #205 PR with your work and a small addition to re-emit these event when the bridge is up.

Can you give it a try ?

Thanks!

**Comment by konradkierus** [comment]
Hi @manuquentin 
Unfortunately, I"m not currently working on any project which is using this approach and I have limited access to iOS device so I won't be able to test your changes.

**Comment by sboily** [comment]
Can we close it and continue the discussion on #205?

**Comment by manuquentin** [comment]
Closed in favor of #205

**Pull request closed without merging**